src/java/org/apache/cassandra/io/sstable/SSTable.java

```diff
@@ -114,6 +119,17 @@ public ImmutableSet<Component> components()
         return components;
     }
 
+    public Optional<CompactionMetadata> getCompactionMetadata() throws IOException
+    {
+        // if compaction metadata is not loaded yet, load it
+        if (compactionMetadata == null)
+        {
+            logger.debug("Loading compaction metadata for {}", descriptor);
+            compactionMetadata = Optional.ofNullable((CompactionMetadata) descriptor.getMetadataSerializer().deserialize(descriptor, MetadataType.COMPACTION));
+        }
```

Comment:
we can use double-check locking to avoid race:
```
    public Optional<CompactionMetadata> getCompactionMetadata() throws IOException
    {
        // if compaction metadata is not loaded yet, load it
        if (compactionMetadata == null)
        {
            synchronized (this)
            {
                if (compactionMetadata == null)
                {
                    logger.debug("Loading compaction metadata for {}", descriptor);
                    compactionMetadata = Optional.ofNullable((CompactionMetadata) descriptor.getMetadataSerializer().deserialize(descriptor, MetadataType.COMPACTION));
                }
            }
        }
        return compactionMetadata;
    }
```

Reply:
The race is benign, I don't think it is worth to make the code more complex.
The worse case is that 2 threads reads the same value and eventually one wins the race